nat-lab/tests/telio.py

```diff
@@ -1048,6 +1050,76 @@ async def get_log(self) -> str:
         await process.execute()
         return process.get_stdout()
 
+    async def get_system_log(self) -> Optional[str]:
+        """
+        Get the system log on the target machine
+        Windows only for now
+        """
+        if self._connection.target_os == TargetOS.Windows:
+            output_notifier = OutputNotifier()
+            done_event = asyncio.Event()
+            done_event_str = "Get-EventLog - Done!"
+
+            async def on_stdout(stdout: str) -> None:
+                for line in stdout.splitlines():
+                    output_notifier.handle_output(line)
+
+            output_notifier.notify_output(
+                f"{done_event_str}",
+                done_event,
+            )
+
+            process = self._connection.create_process([
+                "powershell",
+            ])
+            async with process.run(
+                stdout_callback=on_stdout, stderr_callback=on_stdout
+            ):
+                await process.wait_stdin_ready()
+                await asyncio.sleep(1)
+                await process.write_stdin(
+                    "Get-EventLog -LogName Application -Newest 100 | format-table -wrap\n"
+                )
+                await process.write_stdin(
+                    "Get-EventLog -LogName System -Newest 100 | format-table -wrap\n"
+                )
+                await process.write_stdin(f"echo '{done_event_str}'\n")
+                await done_event.wait()
+                return process.get_stdout()
+        return None
+
+    async def clear_system_log(self) -> None:
+        """
+        Clear the system log on the target machine
+        Windows only for now
+        """
+        if self._connection.target_os == TargetOS.Windows:
+            output_notifier = OutputNotifier()
+            done_event = asyncio.Event()
+            done_event_str = "Clear-EventLog - Done!"
+
+            async def on_stdout(stdout: str) -> None:
+                for line in stdout.splitlines():
+                    output_notifier.handle_output(line)
```

Comment:
Is this really needed here? Can't `output_notifier.handle_output` just be passed as a callback?

Reply:
`output_notifier.handle_output` kinda expects line instead of multiple lines and while `ssh_process` executes callbacks on line, `docker_process` doesn't. I think it is better this way, to be future proof.

Reply:
But why does `output_notifier.handle_output` expect line? I looked at the code and it seems that it should handle multiline input as well. We have multiple places where this thin `on_stdout` wrapper is used and most likely those could be removed. It may be done separately as a small cleanup task. But maybe I'm missing sth

Reply:
We could easily provide multiple lines to output notifier, just variable naming in in it's code suggest, that we expect line, no other constrains :smile: Probably we could easily get rid of those wrappers, will create an issue.

Reply:
LLT-5348